financial-services-accelerator/components/org.wso2.financial.services.accelerator.gateway/src/main/java/org/wso2/financial/services/accelerator/gateway/executor/impl/dcr/DCRExecutor.java

```diff
@@ -0,0 +1,217 @@
+                                    "Provided jwt is malformed and cannot be decoded");
+                            return;
+                        }
+
+                        JSONObject decodedRequestObj = new JSONObject(decodedRequest);
+                        JSONObject decodedSSA;
+                        //Check whether the SSA exists and decode the SSA
+                        if (decodedRequestObj.has(GatewayConstants.SOFTWARE_STATEMENT) &&
+                                decodedRequestObj.getString(GatewayConstants.SOFTWARE_STATEMENT) != null) {
+                            String ssa = JWTUtils.decodeRequestJWT(decodedRequestObj
+                                    .getString(GatewayConstants.SOFTWARE_STATEMENT), "body");
+                            decodedSSA = new JSONObject(ssa);
+                        } else {
+                            handleBadRequestError(fsapiRequestContext,
+                                    "Required parameter software statement cannot be null");
+                            return;
+                        }
+                        JWTClaimsSet requestClaims = GatewayUtils.validateRequestSignature(payload, decodedSSA);
+                        String isDcrPayload = GatewayUtils.constructIsDcrPayload(requestClaims, decodedSSA);
+                        fsapiRequestContext.addContextProperty(GatewayConstants.REQUEST_PAYLOAD,
+                                String.valueOf(new JSONObject(requestClaims.getClaims())));
+
+                        fsapiRequestContext.setModifiedPayload(isDcrPayload);
+                        Map<String, String> requestHeaders = fsapiRequestContext.getMsgInfo().getHeaders();
+                        requestHeaders.remove(GatewayConstants.CONTENT_TYPE_TAG);
+                        Map<String, String> addedHeaders = fsapiRequestContext.getAddedHeaders();
+                        addedHeaders.put(GatewayConstants.CONTENT_TYPE_TAG, GatewayConstants.JSON_CONTENT_TYPE);
+                        fsapiRequestContext.setAddedHeaders(addedHeaders);
+                        fsapiRequestContext.getMsgInfo().setHeaders(requestHeaders);
+                    } else {
+                        handleBadRequestError(fsapiRequestContext, "Malformed request found");
+                    }
+                }
+            } catch (ParseException e) {
+                log.error("Error occurred while decoding the provided jwt", e);
+                handleBadRequestError(fsapiRequestContext, "Malformed request JWT");
+            } catch (BadJOSEException e) {
+                log.error("Error occurred while validating the signature", e);
+                handleBadRequestError(fsapiRequestContext, "Invalid request signature. " + e.getMessage());
+            } catch (JOSEException | MalformedURLException e) {
+                log.error("Error occurred while validating the signature", e);
+                handleBadRequestError(fsapiRequestContext, "Invalid request signature");
```

Comment:
Shall we combine these two exceptions since they handle the same error?

Reply:
In addition, from where can the ```MalformedURLException``` can be thrown and how it relates to signature validation?